Use custom StringWriter to fix encoding.

333fred opened the pull request:
**Customer scenario**

When the customer edits a project file, we currently strip the UTF-16 header as a workaround for https://github.com/dotnet/roslyn-project-system/issues/1168. This is a bad experience, as if the user does want to add an xml header we will strip it away. This fixes the root cause of the issue by creating an override for `StringWriter` which uses the project file encoding.

**Bugs this fixes:**

Fully fixes https://github.com/dotnet/roslyn-project-system/issues/1168.

**Workarounds, if any**

If the user wants to have an XML header, they need to use an external header.

**Risk**

Should be low. MSBuild is using a similar workaround, except that they pass in an encoding in the constructor, instead of reading it from the project.

**Performance impact**

Should be minimal. We're making one more async call, but it should be very quick and we cache the result.

**Is this a regression from a previous update?**

Not a regression.

**Root cause analysis:**

N/A.

**How was the bug found?**

Internal dogfooding and MVP reports.

333fred commented:
Tagging @dotnet/project-system for review.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/MSBuildXmlAccessor.cs:
```diff
@@ -14,33 +12,33 @@ internal class MSBuildXmlAccessor : IProjectXmlAccessor
-                    xmlString = xmlString.Substring(xmlString.IndexOf(Environment.NewLine) + Environment.NewLine.Length);
-                }
-                return xmlString;
+                return stringWriter.ToString();
```
You are leaking the export here, you have to dispose the result of CreateExport

333fred replied:
No longer using CreateExport.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/ProjectEncodingStringWriter.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Editor
+{
+    [Export]
+    internal class ProjectEncodingStringWriter : StringWriter
```
This feels overkill, any reason this just wasn't a custom object non-export that took an Encoding that you passed to it? You can avoid the ExecuteSync because you are in a async method.

davkean commented:
Can you please test the following scenarios that we do not remove/change the encodings:

- I have an XML prolog with UTF-8, no BOM.
- I have an XML prolog with UTF-16, no BOM
- I have an XML prolog with UTF-8 with a UTF-8 BOM
- I have no XML prolog with UTF-8 BOM
- I have no XML prolog with UTF-16 BOM

333fred commented:
@davkean addressed your comments. For the scenarios, it looks like we're preserving the prolog existing or not existing, but always adding a BOM, regardless of whether or not it exists. @jviau and @cdmihai, CPS has this same behavior. If I add a NuGet package from the UI and save the file, without opening the editor, a BOM is added.

davkean commented:
Okay, so sounds like we should file another bug for this.

srivatsn commented:
Tagging @MattGertz for approval.

MattGertz commented:
@srivatsn Let's take to Shiproom, and be prepared to discuss how common the scenario would be.

natidea commented:
Approved to merge (via [370315](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/370315))
